x509util/extensions.go

```diff
@@ -157,6 +189,158 @@ func (s SubjectAlternativeName) Set(c *x509.Certificate) {
+			return SubjectAlternativeName{Type: URIType, Value: s.Value}.RawValue()
+		} else if strings.Contains(s.Value, "@") {
+			return SubjectAlternativeName{Type: EmailType, Value: s.Value}.RawValue()
+		} else {
+			return SubjectAlternativeName{Type: DNSType, Value: s.Value}.RawValue()
+		}
+	case EmailType:
+		valid := isIA5String(s.Value)
+		if !valid {
+			return nil, fmt.Errorf("error converting SAN type %s value to ia5", s.Type)
+		}
+		asnValue = asn1.RawValue{Tag: nameTypeEmail, Class: asn1.ClassContextSpecific, Bytes: []byte(s.Value)}
+	case DNSType:
+		ia5String, err := SanitizeName(s.Value) // use SanitizeName for DNS types because it will do some character replacement and verify that its an acceptable hostname
+		if err != nil {
+			return nil, errors.Wrapf(err, "error converting SAN type %s value to ia5", s.Type)
+		}
+		asnValue = asn1.RawValue{Tag: nameTypeDNS, Class: asn1.ClassContextSpecific, Bytes: []byte(ia5String)}
+	case X400AddressType, DirectoryNameType, EDIPartyNameType:
+		return nil, fmt.Errorf("unimplemented SAN type %s", s.Type)
+	case URIType:
+		valid := isIA5String(s.Value)
+		if !valid {
+			return nil, fmt.Errorf("error converting SAN type %s value to ia5", s.Type)
+		}
+		asnValue = asn1.RawValue{Tag: nameTypeURI, Class: asn1.ClassContextSpecific, Bytes: []byte(s.Value)}
+	case IPType:
+		if rawIP := net.ParseIP(s.Value); rawIP != nil {
+			ip := rawIP.To4()
+			if ip == nil {
+				ip = rawIP
+			}
+			asnValue = asn1.RawValue{Tag: nameTypeIP, Class: asn1.ClassContextSpecific, Bytes: ip}
+		} else {
+			return nil, fmt.Errorf("cannot convert %s to IP", s.Value)
+		}
+	case RegisteredIDType:
+		if s.Value != "" {
+			oid, err := parseObjectIdentifier(s.Value)
+			if err == nil {
+				rawBytes, err := asn1.MarshalWithParams(oid, fmt.Sprintf("tag:%d", nameTypeRegisteredID))
+				if err != nil {
+					return nil, errors.Wrap(err, "unable to Marshal RegisteredID SAN")
+				}
+				asnValue = asn1.RawValue{FullBytes: rawBytes}
+			} else {
+				return nil, errors.Wrap(err, "error parsing OID for RegisteredID SAN")
+			}
+		} else {
+			return nil, errors.New("blank RegisteredID SAN is not allowed")
+		}
+	default:
+		// if san.Type is a valid OID, we assume it is an OtherName
+		oid, err := parseObjectIdentifier(s.Type)
+		if err == nil {
+			// the OtherName value can be any type depending on the OID
+			// ASN supports a great number of formats (https://www.openssl.org/docs/man1.0.2/man3/ASN1_generate_nconf.html),
+			// but golang's asn1 lib supports much fewer -- for now support anything the golang asn1 marshaller supports
+
+			// The default type is printable, but if the value is prefixed with a type, use that
+			var valueType = "printable"
+			var sanValue = s.Value
+			var rawBytes []byte
+			var err error
+
+			if strings.Contains(s.Value, sanTypeSeparator) {
+				valueType = strings.Split(s.Value, sanTypeSeparator)[0]
+				sanValue = s.Value[len(valueType)+1:]
+			}
```

Comment:
@unreality is there any reason to use `;` instead of `:` as the separator? You're linking there to the OpenSSL docs that use `[modifier,]type[:value]`.

Reply:
Not that I can recall off the top of my head...colon is probably fine to use